### PR TITLE
8338688: Shenandoah: Avoid calling java_lang_Class accessors in asserts/verifier

### PR DESCRIPTION
**Pull request by shipilev**
In GC verification code, we are not always safe to touch the klass directly. This becomes a problem in Lilliput, where loading klass from the from-space is erroneous. Lilliput would replace `obj->klass()` with `obj->forward_safe_klass()` to make it right in GC code. But accessors like `java_lang_Class` would not be fixed: they would instead rely on barriers to always be called on to-space objects.

So we are better avoiding using these `java_lang_Class` accessors in GC verification code, and use the loaded `klass` directly. 

Additional tests:
 - [x] Linux x86_64 server fastdebug, `hotspot_gc_shenandoah`
 - [x] Linux x86_64 server fastdebug, `all` with `-XX:+UseShenandoahGC`

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8338688](https://bugs.openjdk.org/browse/JDK-8338688): Shenandoah: Avoid calling java_lang_Class accessors in asserts/verifier (**Enhancement** - P4)

### Reviewers
 * [Roman Kennke](https://openjdk.org/census#rkennke) (@rkennke - **Reviewer**)
 * [William Kemper](https://openjdk.org/census#wkemper) (@earthling-amzn - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20651/head:pull/20651` \
`$ git checkout pull/20651`

Update a local copy of the PR: \
`$ git checkout pull/20651` \
`$ git pull https://git.openjdk.org/jdk.git pull/20651/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20651`

View PR using the GUI difftool: \
`$ git pr show -t 20651`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20651.diff">https://git.openjdk.org/jdk/pull/20651.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20651#issuecomment-2299340291)

### CODE REVIEW

**Review by rkennke — APPROVED** [review]
Looks good to me, and I verified that it fixes the problems that I've observed in Lilliput.

**Comment by shipilev** [comment]
Thanks! I think I need a second review, @earthling-amzn ?

**Review by earthling-amzn — APPROVED** [review]
Thanks, LGTM.

**Comment by shipilev** [comment]
Thank you!

/integrate

**Pull request closed without merging**
